### PR TITLE
add lake checkout command and rework CLI args to use HEAD

### PR DESCRIPTION
**Pull request by mccanne**
This commit adds some more fit and finish to our adaptation of the
Git design pattern.  We added a checkout command and the notion
of HEAD that is tracked in ./.zed_head.  The -p flag has been
eliminated.

The checkout command and the HEAD file provide a convenience and
is not required as all commands have a "-head" option at the lake
level.  We envisage that checkout would be used in interactive
workflows and automation and orchestration tooling would hardwire
their use of -head.

You can now query a lake without an implicit "from" operator,
in which case, HEAD is assumed.  You can also use "from HEAD".

We also changed the output flag -d to -D to make room for the -d
flag in the branch command.

The lakeio printer has been updated to output the familiar HEAD
notation in the log (e.g., "HEAD -> branch").

Finally, we created a new package called lakeparse to hold the
lake-oriented parsing functions that are used by the compiler,
the CLI, and the lake.  ParseID() has finally found its true home.

With a little more massaging, we should be able to checkout a 
detached HEAD with an arbitrary commit ID.  See issue #3019

### CODE REVIEW

**Comment by philrz** [comment]
I started going through the design doc as a user and got stuck at an early step. I attempted:

```
$ zed lake init
using environment variable ZED_LAKE_ROOT
lake created: /Users/phil/logs

$ zed lake create -orderby ts:desc logs
pool created: logs 1xlw421Hz1t4zCDceK82ihaaP7J

$ zed lake load *
HEAD not specified: run checkout or use -HEAD
```

Which I thought to try because of this text:

> Zed lakes can be used without thinking about branches. When referencing a pool without a branch, the tooling presumes the "main" branch as the default, and everything can be done on main without having to think about branching.

If that text still is still presumed accurate, what's the way that I'd now reference a pool without a branch in the `load` command? The only way I can see to do that is to run `zed lake checkout logs@main` after I create the pool and before I attempt the `load`, but that has me (as a user) thinking about branches. Maybe an automatic checkout of `main` of the created pool is appropriate as a side effect of the `create`? This would seem similar how when I do a `git clone` it automatically checks out the default branch.

Separate topic: I also see `-p pool` is still mentioned in lots of the CLI help.

File: cmd/zed/lake/branch/command.go [hunk 1]
```diff
@@ -38,46 +45,56 @@ func init() {
...
-	poolName, commitName := c.lakeFlags.Branch()
-	if poolName == "" {
-		return errors.New("a branch/commit must be specified with -p")
+	if len(args) == 0 {
+		return c.list(ctx, lake)
 	}
-	lake, err := c.lake.Open(ctx)
+	if len(args) > 1 {
+		return errors.New("too many arguments")
+	}
```
Comment (nwt) [inline cmd/zed/lake/branch/command.go:78]: Maybe do this before opening the lake?

File: cmd/zed/lake/branch/command.go [hunk 2]
```diff
@@ -99,3 +116,27 @@ func (c *Command) Run(args []string) error {
...
 }
+
+func (c *Command) list(ctx context.Context, lake api.Interface) error {
+	head, err := c.lakeFlags.HEAD()
+	if err != nil {
+		return err
+	}
+	poolName := head.Pool
+	if poolName == "" {
+		return errors.New("must be on a checked out out branch to list the branches in the same pool")
```
Comment (nwt) [inline cmd/zed/lake/branch/command.go:127]:
```suggestion
		return errors.New("must be on a checked out branch to list the branches in the same pool")
```

File: cmd/zed/lake/checkout/command.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	"github.com/brimdata/zed/pkg/charm"
+)
+
+var Checkout = &charm.Spec{
+	Name:  "checkout",
+	Usage: "checkout [-b] branch",
+	Short: "checkout a branch",
+	Long: `
+The lake checkout command sets the working branch as indicated.
+This allows commands like load, rebase, merge etc to function without
```
Comment (nwt) [inline cmd/zed/lake/checkout/command.go:21]:
```suggestion
This allows commands like load, rebase, merge, etc. to function without
```

File: cmd/zed/lake/create/command.go [hunk 4]
```diff
@@ -16,15 +16,15 @@ import (
 
 var Create = &charm.Spec{
 	Name:  "create",
-	Usage: "create [-orderby key[,key...][:asc|:desc]] -p name",
+	Usage: "create [-orderby key[,key...][:asc|:desc]] name",
 	Short: "create a new data pool",
 	Long: `
 The lake create command creates new pools.  One or more pool keys may be specified
 as the sort keys (primary, secondary, etc) of the data stored in the pool.
```
Comment (nwt) [inline cmd/zed/lake/create/command.go:23]:
```suggestion
as the sort keys (primary, secondary, etc.) of the data stored in the pool.
```

File: cmd/zed/lake/checkout/command.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+having to specify the working branch.  The branch specifier may also be
+a commit ID, in which case you entered a headless state and commands
```
Comment (nwt) [inline cmd/zed/lake/checkout/command.go:23]:
```suggestion
a commit ID, in which case you will enter a headless state and commands
```

File: cmd/zed/lake/checkout/command.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+like load that require a branch name for HEAD will report an error.
+
+Any command that relies upon HEAD can also be run with the -HEAD option
+to refer to a different HEAD without performing a checkout.
+The HEAD option has the form "pool@branch" where pool is the name or ID of an
```
Comment (nwt) [inline cmd/zed/lake/checkout/command.go:28]:
```suggestion
The HEAD option has the form "pool@commitsh" where pool is the name or ID of an
```

File: cmd/zed/lake/checkout/command.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+existing pool and branch is the name of the branch or a commit ID.
```
Comment (nwt) [inline cmd/zed/lake/checkout/command.go:29]:
```suggestion
existing pool and branch is the name of a branch or a commit ID.
```

File: cmd/zed/lake/checkout/command.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+While checkouts of HEAD are useful for interactive CLI sessions,
+automation and orchestration tools are better of hard-wiring the
```
Comment (nwt) [inline cmd/zed/lake/checkout/command.go:31]:
```suggestion
automation and orchestration tools are better off hard-wiring the
```

File: cmd/zed/lake/checkout/command.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+HEAD references in each lake command using -HEAD.
+
+If the -b option is provided, then a new branch with the indicated
+name is created with base HEAD and checked out.
+
+The checkout command merely checks that the branch exists and updates the
+file ~/.zed_head.  This file simply contains a pointer to the HEAD branch
```
Comment (nwt) [inline cmd/zed/lake/checkout/command.go:38]:
```suggestion
file .zed_head in the current directory.  This file simply contains a pointer to the HEAD branch
```

File: cmd/zed/lake/checkout/command.go [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+	} else if _, err = lake.CommitObject(ctx, poolID, branchName); err != nil {
+		return err
+	}
+	if err := lakeflags.WriteHead(poolName, branchName); err != nil {
+		return err
+	}
+	if !c.lakeFlags.Quiet {
+		new := ""
+		if c.branch {
+			new = "a new "
```
Comment (nwt) [inline cmd/zed/lake/checkout/command.go:133]:
```suggestion
			new = "new "
```

File: cmd/zed/lake/delete/command.go [hunk 11]
```diff
@@ -8,17 +8,18 @@ import (
...
-	Usage: "delete -p pool[@commit] id [id ...]",
-	Short: "delete commits or data objects from a pool branch",
+	Usage: "delete id [id ...]",
+	Short: "delete data objects from a pool branch",
 	Long: `
-"zed lake delete" takes a list of commit tags and/or data object tags
-in the specified pool branch and deletes all of the corresponding data objects.
+The delete command takes a list of data object IDs and
+deletes references to those object from HEAD by commiting a new
+delet operation to HEAD.
```
Comment (nwt) [inline cmd/zed/lake/delete/command.go:22]:
```suggestion
delete operation to HEAD.
```

File: cmd/zed/lake/query/command.go [hunk 12]
```diff
@@ -80,7 +76,8 @@ func (c *Command) Run(args []string) error {
 	if !c.lakeFlags.Quiet {
 		d.SetWarningsWriter(os.Stderr)
 	}
-	stats, err := lake.Query(ctx, d, src, c.includes...)
+	head, _ := c.lakeFlags.HEAD()
```
Comment (nwt) [inline cmd/zed/lake/query/command.go:79]: Handle error?

File: cmd/zed/lake/vacate/command.go [hunk 13]
```diff
@@ -10,22 +10,21 @@ import (
...
+	Usage: "vacate [options] commit",
+	Short: "compact a pool's commit history by squashing old commit objects",
 	Long: `
-The vacate command advances the tail of a pool's commit journal so that any commits
-before the new tail cannot be accessed and thus "time travel" to previous
-such commits can no longer be accomplished.  Data objects that are no
-longer accessible after the tail has been advanced are removed from the
-underlying storage system.
+The vacate command compacts the commit history by squashing all of the commit
+objects in the history up to the indicated commit and removin the old commits.
```
Comment (nwt) [inline cmd/zed/lake/vacate/command.go:17]:
```suggestion
objects in the history up to the indicated commit and removing the old commits.
```

File: cmd/zed/lake/vacate/command.go [hunk 14]
```diff
@@ -10,22 +10,21 @@ import (
...
-The only time you should use
-this is when you want to free up old data that no longer needs to be accessed.
+The branch history may contain pointers to old commit objects, but any attempt
+to access them will fail as the underlying commit history will be no longer available.
 
 DANGER ZONE.
 There is no prompting or second chances here so use carefully.
-Once the pool's tail has been advanced and old data is deleted,
-it cannot be recovered.
+Once the pool's commit history has been sequashed and old commits is deleted,
```
Comment (nwt) [inline cmd/zed/lake/vacate/command.go:26]:
```suggestion
Once the pool's commit history has been squashed and old commits is deleted,
```

File: docs/lake/design.md [hunk 15]
```diff
@@ -262,9 +303,10 @@ every journal entry, which has a Zed type signature as follows:
     Data: <any>
 }
 ```
-None of the fields are used by the Zed lake system for any purpose
-except to provide information about the commit object to the end user
-and/or end application.  Any ZSON/ZNG data can be stored in the `Data` field
+The `Date` field here is used by the Zed lake system to do time travel
+through the branch and pool history, allowing you to see the state of
+branches at their commit history at any point in time.
```
Comment (nwt) [inline docs/lake/design.md:308]:
```suggestion
branches at any time in their commit history.
```

File: lakeparse/head.go [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+	Pool   string
+	Branch string
+}
+
+func ParseCommitish(commitish string) (*Commitish, error) {
+	if strings.IndexByte(commitish, '\'') >= 0 {
+		return nil, errors.New("pool and branch names may not contain single quote characters")
+	}
+	if i := strings.LastIndexByte(commitish, '@'); i > -1 {
+		return &Commitish{commitish[:i], commitish[i+1:]}, nil
```
Comment (nwt) [inline lakeparse/head.go:19]:
```suggestion
		return &Commitish{Pool: commitish[:i], Branch: commitish[i+1:]}, nil
```

File: zio/anyio/writer.go [hunk 17]
```diff
@@ -29,6 +29,7 @@ type WriterOpts struct {
 	Zng    zngio.WriterOpts
 	ZSON   zsonio.WriterOpts
 	Zst    zstio.WriterOpts
+	Lake   lakeio.WriterOpts
```
Comment (nwt) [inline zio/anyio/writer.go:32]: Put this after JSON so the WriterOpts fields remain sorted by name.

**Review by nwt — APPROVED** [review]

**Pull request merged**